packages/crypto/src/Bip32/Bip32KeyDerivation.ts

```diff
@@ -0,0 +1,175 @@
+import { BN } from 'bn.js';
+import { InvalidArgumentError } from '../errors';
+import { crypto_auth_hmacsha512, crypto_scalarmult_ed25519_base_noclamp } from 'libsodium-wrappers-sumo';
+import { eddsa } from 'elliptic';
+
+const ed25519 = new eddsa('ed25519');
+
+/**
+ * Check if the index is hardened.
+ *
+ * @param index The index to verify.
+ * @returns true if hardened; otherwise; false.
+ */
+const isHardenedDerivation = (index: number) => index >= 0x80_00_00_00;
+
+/**
+ * Derives the private key with a hardened index.
+ *
+ * @param index The derivation index.
+ * @param scalar Ed25519 curve scalar.
+ * @param iv Ed25519 binary blob used as IV for signing.
+ * @param chainCode The chain code.
+ */
+const deriveHardened = (
+  index: number,
+  scalar: Buffer,
+  iv: Buffer,
+  chainCode: Buffer
+): { zMac: Uint8Array; ccMac: Uint8Array } => {
+  const data = Buffer.allocUnsafe(1 + 64 + 4);
+  data.writeUInt32LE(index, 1 + 64);
+  scalar.copy(data, 1);
+  iv.copy(data, 1 + 32);
+
+  data[0] = 0x00;
+  const zMac = crypto_auth_hmacsha512(data, chainCode);
+  data[0] = 0x01;
+  const ccMac = crypto_auth_hmacsha512(data, chainCode);
+
+  return { ccMac, zMac };
+};
+
+/**
+ * Derives the private key with a 'soft' index.
+ *
+ * @param index The derivation index.
+ * @param scalar Ed25519 curve scalar.
+ * @param chainCode The chain code.
+ */
+const deriveSoft = (index: number, scalar: Buffer, chainCode: Buffer): { zMac: Uint8Array; ccMac: Uint8Array } => {
+  const data = Buffer.allocUnsafe(1 + 32 + 4);
+  data.writeUInt32LE(index, 1 + 32);
+
+  const vk = Buffer.from(crypto_scalarmult_ed25519_base_noclamp(scalar));
+
+  vk.copy(data, 1);
+
+  data[0] = 0x02;
+  const zMac = crypto_auth_hmacsha512(data, chainCode);
+  data[0] = 0x03;
+  const ccMac = crypto_auth_hmacsha512(data, chainCode);
+
+  return { ccMac, zMac };
+};
+
+/**
+ * Adds the left hand side to 28 bytes of the right hand side and multiplies the result by 8.
+ *
+ * @param lhs Left hand side Little-Endian big number.
+ * @param rhs Right hand side Little-Endian big number.
+ */
+const truc28Mul8 = (lhs: Uint8Array, rhs: Uint8Array): Buffer =>
+  new BN(lhs, 16, 'le').add(new BN(rhs.slice(0, 28), 16, 'le').mul(new BN(8))).toArrayLike(Buffer, 'le', 32);
+
+/**
+ * Computes `(8 * sk[:28])*G` where `sk` is a little-endian encoded int and `G` is the curve's base point.
+ *
+ * @param sk The secret key.
+ */
+const pointOfTrunc28Mul8 = (sk: Uint8Array) => {
+  const left = new BN(sk.slice(0, 28), 16, 'le').mul(new BN(8));
+  return ed25519.curve.g.mul(left);
```

Comment:
Out of curiosity, why are we using `ed25519.curve.g.mul` here instead of `crypto_scalarmult_ed25519_base_noclamp` as used in the ed25519 implementation?

Reply:
I'm not sure I fully understood the problem. You could also use `crypto_core_ed25519_add` from libsodium for the arithmetic. And what format are you trying to decode the point to?